test/shared/src/main/scala/zio/test/package.scala

```diff
@@ -1013,174 +1015,56 @@ package object test extends CompileVariants {
-    ) extends State
-
-    def pull(
-      outerDone: Ref[Boolean],
-      outerStream: ZIO[R, Option[Nothing], Chunk[Option[A]]],
-      currentOuterChunk: Ref[(Chunk[Option[A]], Int)],
-      currentInnerStream: Ref[Option[PullInner]],
-      currentStreams: Ref[ScalaQueue[State]],
-      innerFinalizers: ZManaged.ReleaseMap
-    ): ZIO[R1, Option[Nothing], Chunk[Option[B]]] = {
-
-      def pullNonEmpty[R, E, A](pull: ZIO[R, Option[E], Chunk[A]]): ZIO[R, Option[E], Chunk[A]] =
-        pull.flatMap(as => if (as.nonEmpty) ZIO.succeed(as) else pullNonEmpty(pull))
-
-      def pullOuter(
-        outerStream: ZIO[R, Option[Nothing], Chunk[Option[A]]],
-        outerChunk: Chunk[Option[A]],
-        outerChunkIndex: Int
-      ): ZIO[R1, Option[Nothing], (Option[A], Chunk[Option[A]], Int)] =
-        if (outerChunkIndex < outerChunk.size)
-          ZIO.succeedNow((outerChunk(outerChunkIndex), outerChunk, outerChunkIndex + 1))
-        else
-          pullNonEmpty(outerStream).map(chunk => (chunk(0), chunk, 1))
-
-      def openInner(a: A): ZIO[R1, Nothing, (ZIO[R1, Option[Nothing], Chunk[Option[B]]], ZManaged.Finalizer)] =
-        ZIO.uninterruptibleMask { restore =>
-          for {
-            releaseMap <- ZManaged.ReleaseMap.make
-            pull       <- restore(ZManaged.currentReleaseMap.locally(releaseMap)(f(a).toPull.zio).map(_._2))
-            finalizer  <- innerFinalizers.add(releaseMap.releaseAll(_, ExecutionStrategy.Sequential))
-          } yield (pull, finalizer)
-        }
-
-      def pullInner(
-        innerStream: ZIO[R1, Option[Nothing], Chunk[Option[B]]],
-        innerChunk: Chunk[Option[B]],
-        innerChunkIndex: Int
-      ): ZIO[R1, Option[Nothing], (Option[Chunk[Option[B]]], Chunk[Option[B]], Int)] =
-        if (innerChunkIndex < innerChunk.size)
-          ZIO.succeedNow(takeInner(innerChunk, innerChunkIndex))
-        else
-          pullNonEmpty(innerStream).map(takeInner(_, 0))
-
-      def takeInner(
-        innerChunk: Chunk[Option[B]],
-        innerChunkIndex: Int
-      ): (Option[Chunk[Option[B]]], Chunk[Option[B]], Int) =
-        if (innerChunk(innerChunkIndex).isEmpty) {
-          (None, innerChunk, innerChunkIndex + 1)
-        } else {
-          val builder  = ChunkBuilder.make[Option[B]]()
-          val length   = innerChunk.length
-          var continue = true
-          var i        = innerChunkIndex
-          while (continue && i != length) {
-            val b = innerChunk(i)
-            if (b.isDefined) {
-              builder += b
-              i += 1
-            } else {
-              continue = false
-            }
-          }
-          (Some(builder.result()), innerChunk, i)
-        }
-
-      currentInnerStream.get.flatMap {
-        case None =>
-          currentStreams.get.map(_.headOption).flatMap {
+  )(f: A => ZStream[R1, Nothing, Option[B]])(implicit trace: ZTraceElement): ZStream[R1, Nothing, Option[B]] =
+    new ZStream(
+      stream
+        .rechunk(1)
```

Comment:
Is this necessary?

Reply:
Currently the "customization" only allows to switch streams after each emit, so if there would be a `None` in the middle of the chunk we could not split it and only emit the first half and store the rest for a next run.

For that we would have to add the ability to also switch where we are reading from, to be able to continue from a buffered remainder of the chunk, and also probably to make the `onEmit` an effect.

Reply:
Maybe we could also "dynamically" chunk it at each `None`?

Reply:
Sure.